src/kernel/init.c

```diff
@@ -75,9 +75,10 @@ void init_kernel_heaps(void)
 #endif
     assert(heaps.linear_backed != INVALID_ADDRESS);
 
-    bytes pagesize = is_low_memory_machine(&heaps) ?
-                     U64_FROM_BIT(MAX_MCACHE_ORDER + 1) : PAGESIZE_2M;
-    heaps.general = allocate_mcache(&bootstrap, (heap)heaps.linear_backed, 5, MAX_MCACHE_ORDER,
+    boolean is_lowmem = is_low_memory_machine();
+    int max_mcache_order = is_lowmem ? MAX_LOWMEM_MCACHE_ORDER : MAX_MCACHE_ORDER;
+    bytes pagesize = is_lowmem ? U64_FROM_BIT(max_mcache_order + 1) : PAGESIZE_2M;
+    heaps.general = allocate_mcache(&bootstrap, (heap)heaps.linear_backed, 5, max_mcache_order,
```

Comment:
The general heap is used by the tls klib for its malloc/free implementation; if I try to use this klib in a a low-memory VM, when mbedtls_calloc() in klib/mbedtls.c is called to allocate memory chunks larger than 2KB, memory is allocated directly from the linear backed heap, which causes a subsequent deallocation attempt to fail with something like "assertion f->avail <= o->objs_per_page failed at /usr/src/nanos/src/runtime/heap/objcache.c:176  in objcache_deallocate(); halt".
Since the tls klib can't be used if it can't allocate memory chunks larger than 2 KB, we could prevent it from being loaded in the first place when running in low-memory VMs, e.g. by adding a check in its init() function. But this would prevent us from using the cloud_init klib even if we don't need HTTPS. So we should probably add a check in tls_connect() instead, so that the klib loads correctly but returns an error when one tries to set up a TLS session.

Reply:
I have addressed this issue by tracking such "fallback" allocations in a hash table stored within the mcache struct. When an allocation is made that exceeds the size of the largest contained mcache, the size is stored in said table with the address as the key. When deallocating objects of unknown (-1ull) size, a lookup/removal is made in this table and, if an entry is found, the size parameter is padded and used for the deallocate. Small objects are not recorded, so a lookup failure proceeds to bin the allocation according to objcache page membership as was done prior. Deallocation of objects with a specified size classified as fallback also requires removal from the table. The cost here is a table insertion and removal for all objects of fallback size as well as a lookup/removal for all deallocations of an unspecified size. Given that such uses are not typical during kernel runtime and are primarily for vendor code use, this should add no cost in most cases, and even a negligible one for said exceptions.

The upshot here is that the mcache may now be used with a "malloc/free"-type interface regardless of allocation size. We don't have an exhaustive mcache test (#1886), but I have observed such fallback deallocations working correctly with a test run using cloud-init.